reservation_units/querysets.py

```diff
@@ -0,0 +1,418 @@
+
+        Date and Time filters are used to filter a range of dates and time when the reservation must be within.
+
+        This method works by first generating a list of closed time spans for each ReservationUnit and then one by one
+        checking the reservable time spans for a time span that is not overlapping with closed time spans and still
+        long enough to fulfill the minimum duration criteria.
+
+        The list of closed time spans are generated from:
+        - Given filter value
+        - ReservationUnit settings
+        - ReservationUnit's ApplicationRounds
+        - Reservations from ReservationUnit's TILAHIERARKIA TODO
+        After removing the closed time spans, the first reservable time span that is long enough to fit the minimum
+        duration is selected as the `first_reservable_datetime`.
+
+        Variables which affect the first reservable time span and cause the ReservationUnit to be "closed":
+        - Date filter range
+        - Time filter range
+        - ReservationUnit `reservation_begins`
+        - ReservationUnit `reservation_ends`
+        - ReservationUnit `publish_ends`
+        - ApplicationRound `reservation_period_begin`
+        - ApplicationRound `reservation_period_end`
+
+        Variables which only affect the first reservable time span and allow the ReservationUnit to be "open":
+        - Minimum duration filter
+        - ReservationUnit `min_reservation_duration`
+        - ReservationUnit `max_reservation_duration`
+        - ReservationUnit `reservation_start_interval`
+        - ReservationUnit `reservations_min_days_before`
+        - ReservationUnit `reservations_max_days_before`
+        - Reservations from ReservationUnit TILAHIERARKIA TODO
+        """
+        now = timezone.localtime()
+        today = now.date()
+        two_years_from_now = today + timedelta(days=730)
+
+        #########################
+        # Default filter values #
+        #########################
+
+        if filter_date_start is None:
+            filter_date_start = today
+        if filter_date_end is None:
+            filter_date_end = two_years_from_now
+
+        if filter_time_start is not None:
+            filter_time_start = filter_time_start.replace(tzinfo=DEFAULT_TIMEZONE)
+        if filter_time_end is not None:
+            filter_time_end = filter_time_end.replace(tzinfo=DEFAULT_TIMEZONE)
+
+        ##########################
+        # Validate filter values #
+        ##########################
+
+        if filter_date_start < today:
+            raise ValueError("'reservable_date_start' must be not be in the past.")
+        elif filter_date_end < today:
+            raise ValueError("'reservable_date_end' must be not be in the past.")
+        elif filter_date_end > two_years_from_now:
+            raise ValueError("'reservable_date_end' must be not be more than two years in the future.")
```

Comment:
If there is a leap year in between, then I guess range like 19.12.2023-19.12.2025 would raise exception?
That could be solved by two_years_from_now = today + timedelta(days=**731**) -- no side effects for time span checks (the last day in that case just is not in the DB so simply wouldn't come up as a first reservable date in the edge case anywhow but the exception would be avoided)?